src/app/decorations/qgsdecorationoverlay.cpp

```diff
@@ -0,0 +1,64 @@
+/***************************************************************************
+                         qgsdecorationoverlay.cpp
+                         ----------------------
+    begin                : July 2023
+    copyright            : (C) 2023 by Yoann Quenach de Quivillic
+    email                : yoann dot quenach at gmail dot com
+
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsdecorationoverlay.h"
+#include "qgsmapdecoration.h"
+#include "qgsmapcanvas.h"
+#include "qgisapp.h"
+
+#include <QPainter>
+
+QgsDecorationOverlay::QgsDecorationOverlay( QWidget *parent )
+  : QWidget( parent )
+{
+  // Make the overlay transparent
+  setAttribute( Qt::WA_NoSystemBackground );
+
+  // Make the overlay transparent for mouse events
+  setAttribute( Qt::WA_TransparentForMouseEvents );
+
+  // Install the event filter to catch resize events
+  parent->installEventFilter( this );
+}
+
+
+void QgsDecorationOverlay::paintEvent( QPaintEvent * )
+{
+  QPainter p( this );
+
+  QgsRenderContext context = QgsRenderContext::fromMapSettings( QgisApp::instance()->mapCanvas()->mapSettings() );
+  context.setPainter( &p );
+
+  for ( QgsMapDecoration *item : QgisApp::instance()->activeDecorations() )
+  {
+    // Do not render decorations with fixed map positionn they are rendered directly on the map canvas
+    if ( item->hasFixedMapPosition() )
+      continue;
+    item->render( QgisApp::instance()->mapCanvas()->mapSettings(), context );
+  }
```

Comment:
Let's save some unnecessary work when no decorations are around:

```suggestion
  const QList< QgsMapDecoration* > decorations = QgisApp::instance()->activeDecorations();
  if ( decorations.empty() )
    return;
    
  QPainter p( this );

  QgsRenderContext context = QgsRenderContext::fromMapSettings( QgisApp::instance()->mapCanvas()->mapSettings() );
  context.setPainter( &p );

  for ( QgsMapDecoration *item : decorations )
  {
    // Do not render decorations with fixed map positionn they are rendered directly on the map canvas
    if ( item->hasFixedMapPosition() )
      continue;
    item->render( QgisApp::instance()->mapCanvas()->mapSettings(), context );
  }
```